File: 프로그래머스/2/12946. 하노이의 탑/하노이의 탑.py
def solution(n):
    answer = []
    
    
    hanoi(n, 1, 3, answer)
    
    return answer

def hanoi(n, a, b, answer):
    if n == 1:
        answer.append([a,b])
        return
    
    hanoi(n-1, a, 6-a-b, answer)
    hanoi(1, a, b, answer)
    hanoi(n-1,6-a-b,b, answer)

File: 프로그래머스/2/12946. 하노이의 탑/하노이의 탑.java
import java.util.*;


class Solution {
    public int[][] solution(int n) {
        List<int[]> answer = new LinkedList<>();
        hanoi(answer, n, 1, 3);
        return answer.toArray(new int[0][]);       
    }
    
    
    public void hanoi(List<int[]> answer, int n, int start, int end) {
        if (n == 0) {
            return;
        }
        
        hanoi(answer, n-1, start, 6-start-end);
        answer.add(new int[]{start, end});
        hanoi(answer, n-1, 6-start-end, end);
    }
}